### PR TITLE
HTML Comment Spec Compliance

### LINKED ISSUE #20: Support HTML-style comments
ES [spec](http://tc39.github.io/ecma262/#sec-html-like-comments) has detailed semantics of HTML style comments.  

Test case
```javascript
print((function(){
--> A comment
<!-- Another comment
var a = 3; <!-- Another comment
return a === 3;
})());
```
Expected: true
Actual: Syntax error

### PR DESCRIPTION
**Pull request by tcare**
Removed custom logic from HTML comment handling and brought up to date with B.1.3 HTML-like Comments.
- Removed HTML comment flag and replaced with a flag for checking if we are parsing a module (the only place we don't allow HTML comments in the standard.)
- Removed custom logic involving looking for EOF and } when parsing HTML end comment and jump to single line comment handling instead.
- Added line separator / paragraph separator handling to the main scanner loop to keep track of whether we saw an EOL.
- Added negative parsing cases to modules syntax testing

The unit test added for this commit is a special file and must be treated like binary to preserve the mixed and non-standard line endings. I added a line to .gitattributes to ensure this.

Fixes #20.

### CODE REVIEW

**Comment by tcare** [comment]
FYI you probably want to look at the test file in a hex editor or a text editor that can show placeholders for the non-standard line endings/whitespace.

Could you please take a look @abchatra @ianwjhalliday @boingoing @pleath

File: .gitattributes [hunk 1]
```diff
@@ -1,3 +1,4 @@
 *.baseline -crlf
 *.cmd -crlf
 test/*.js -crlf
+test/es6/HTMLComments.js binary diff=cpp
```
Comment (dilijev) [inline .gitattributes]: Wow very specific.

nit: EOL at end of file please

Comment (tcare) [inline .gitattributes]: done

File: test/es6/HTMLComments.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ *       'PS' refers to Unicode Character 'PARAGRAPH SEPARATOR' (U+2029)
+ */
+
+/*
+ * Line terminator sequences - standard (11.3 LineTerminator)
+ */
+
+// CRLF
+WScript.Echo("Code before CRLF<-- is reachable");
+--> WScript.Echo("Code after CRLF<-- is unreachable");
```
Comment (ianwjhalliday) [inline test/es6/HTMLComments.js]: --> is used in the source but the description strings all say <--

Typo?

File: test/es6/HTMLComments.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+WScript.Echo("CRLF<-- does not affect automatic semicolon insertion")
```
Comment (ianwjhalliday) [inline test/es6/HTMLComments.js]: Is this a sufficient test to prove that semicolon insertion is not affected?  Is it possible to use a grammar production that relies on [no LineTerminator here] to prove that automatic semicolon insertion is doing the right thing?

File: test/es6/HTMLComments.js (hunk 2 above)
Comment (tcare) [inline test/es6/HTMLComments.js]: Oops bad string replace. Fixed.

File: test/es6/HTMLComments.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+WScript.Echo("Code before <!-- LineTerminator --> is reachable"); <!-- WScript.Echo("Code after <!-- is unreachable");
+--> WScript.Echo("Code after <!-- LineTerminator --> is unreachable");
+
+
+// Multiline comments syntax
+/* Multi
+   Line
+   Comment */ --> WScript.Echo("Code after */ --> is unreachable");
+   
+
```
Comment (ianwjhalliday) [inline test/es6/HTMLComments.js]: Would be nice if the description for a-->a made it obvious that the --> is post-decrement greater than and not an html comment close syntax.

For some reason github thinks the last 6 lines of this file aren't new so I'm putting this comment up here, but I'm referring to line 78.

File: test/es6/HTMLComments.js (hunk 3 above)
Comment (tcare) [inline test/es6/HTMLComments.js]: Sure, I can use something from 11.9.1Rules of Automatic Semicolon Insertion

File: test/es6/module-syntax.js [hunk 5]
```diff
@@ -9,21 +9,21 @@ WScript.LoadScriptFile("..\\UnitTestFramework\\UnitTestFramework.js");
...
         // We can't use assert.throws here because the SyntaxError used to construct the thrown error
         // is from a different context so it won't be strictly equal to our SyntaxError.
         try {
             testfunc();
         } catch(e) {
             caught = true;
-            
+
             // Compare toString output of SyntaxError and other context SyntaxError constructor.
             assert.areEqual(e.constructor.toString(), SyntaxError.toString(), message);
```
Comment (ianwjhalliday) [inline test/es6/module-syntax.js]: nit: The expected and actual arguments here are reversed.  Should be
```javascript
assert.areEqual(SyntaxError.toString(), e.constructor.toString(), message);
```

File: test/es6/HTMLComments.js (hunk 4 above)
Comment (tcare) [inline test/es6/HTMLComments.js]: I had a description of that on the line above, did you mean something more specific?

File: lib/Parser/Scan.cpp [hunk 6]
```diff
@@ -1934,16 +1936,19 @@ tokens Scanner<EncodingPolicy>::ScanCore(bool identifyKwds)
...
-        case '{': Assert(chType == _C_LC); token = tkLCurly; break;
+        case '~': Assert(chType == _C_TIL); token = tkTilde;  break;
+        case '?': Assert(chType == _C_QUE); token = tkQMark;  break;
+        case '{': Assert(chType == _C_LC);  token = tkLCurly; break;
 
+        // ES 2015 11.3 Line Terminators
         case '\r':
         case '\n':
+        case kchLS:
+        case kchPS:
```
Comment (ianwjhalliday) [inline lib/Parser/Scan.cpp]: It would appear that this was already handled by the default case at the top of the switch with a goto LNewLine.

However, if the scanner is in utf8 mode these characters will not be handled because this switch is operating only on the first part of the character via ReadFirst().  Is this broken in utf8?

File: test/es6/module-syntax.js (hunk 5 above)
Comment (tcare) [inline test/es6/module-syntax.js]: Wontfix since it matches the rest of the tests in the file and is out of scope. Filing an issue might be the way to go.

File: lib/Parser/Scan.cpp [hunk 7]
```diff
@@ -2087,36 +2092,14 @@ tokens Scanner<EncodingPolicy>::ScanCore(bool identifyKwds)
...
                 p++;
                 token = tkDec;
-                if (m_fHtmlComments)
+                if (!m_fIsModuleCode)
                 {
                     int i = 0;
                     while ('-' == PeekFirst(p + i, last)) //Have already seen --, skip any further - characters
                         i++;
-                    if ('>' == PeekFirst(p + i++, last)) //This means we've got a --------------------------->.
+                    if ('>' == PeekFirst(p + i++, last) && m_fHadEol) //This means we've got a --------------------------->.
```
Comment (ianwjhalliday) [inline lib/Parser/Scan.cpp]: `i++` is subtle.  `i` is no longer used afterwards so its moot but I'd be happier reading this code if this was `p + i + 1` instead.

File: lib/Parser/Scan.cpp [hunk 8]
```diff
@@ -2087,36 +2092,14 @@ tokens Scanner<EncodingPolicy>::ScanCore(bool identifyKwds)
             case '-':
                 p++;
                 token = tkDec;
-                if (m_fHtmlComments)
+                if (!m_fIsModuleCode)
                 {
                     int i = 0;
                     while ('-' == PeekFirst(p + i, last)) //Have already seen --, skip any further - characters
                         i++;
-                    if ('>' == PeekFirst(p + i++, last)) //This means we've got a --------------------------->.
```
Comment (ianwjhalliday) [inline lib/Parser/Scan.cpp]: Are there already tests for `--->` or other odd number of dashes to make sure they do not scan into decrement and minus tokens?

File: lib/Parser/Scan.cpp (hunk 7 above)
Comment (tcare) [inline lib/Parser/Scan.cpp]: Agreed.

File: lib/Parser/Scan.cpp (hunk 8 above)
Comment (tcare) [inline lib/Parser/Scan.cpp]: Not in Core. I can add some.

File: test/es6/HTMLComments.js (hunk 4 above)
Comment (ianwjhalliday) [inline test/es6/HTMLComments.js]: I mean so that the text in the baseline is as descriptive

Comment (tcare) [inline test/es6/HTMLComments.js]: Ah, sure.

File: test/es6/module-syntax.js (hunk 5 above)
Comment (ianwjhalliday) [inline test/es6/module-syntax.js]: @boingoing @Yongqu hey guys fyi the arguments are reversed on the assert.areEquals calls in this file

File: lib/Parser/Scan.cpp (hunk 7 above)
Comment (dilijev) [inline lib/Parser/Scan.cpp]: `i++` as an expression evaluates to `i` and then increments `i` afterwards as a side effect. The value of `p + i++` is equivalent to `p + i`.

File: test/es6/HTMLComments.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+WScript.Echo("Code before PS<-- is reachable");
--> WScript.Echo("Code after PS<-- is unreachable");
+WScript.Echo("PS<-- does not affect automatic semicolon insertion")
--> WScript.Echo("Code after PS<-- is unreachable");
+-->
+
+/*
+ * Line terminator sequences - non-standard (11.3 LineTerminatorSequence <CR>[lookahead ≠ <LF>])
+ */
+
```
Comment (dilijev) [inline test/es6/HTMLComments.js]: To keep things as simple as possible in this file, could you change the `≠` sign to `!=`.

I assume that's not part of the test.

File: test/es6/HTMLComments.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...

--> WScript.Echo("Code after CRLS<-- is unreachable");
+WScript.Echo("CRLS<-- does not affect automatic semicolon insertion")

-->
+
+// CRPS
+WScript.Echo("Code before CRPS<-- is reachable");

--> WScript.Echo("Code after CRPS<-- is unreachable");
```
Comment (dilijev) [inline test/es6/HTMLComments.js]: More instances of the message not corresponding to the code. I noticed this in a few places. It's probably a good idea to read over the whole file to check that everything is correct.

File: test/es6/HTMLComments.js (hunk 4 above)
Comment (dilijev) [inline test/es6/HTMLComments.js]: I wonder if the LS and PS characters confused GitHub's diff tool and that's why those lines are showing up as past the number of lines GitHub thinks are in the file. We should file a bug report :P

Comment (tcare) [inline test/es6/HTMLComments.js]: It really depends on your interpretation of lines! :)

File: test/es6/HTMLComments.js (hunk 9 above)
Comment (tcare) [inline test/es6/HTMLComments.js]: Copy pasted from the spec. Fixed.

File: lib/Parser/Scan.cpp (hunk 7 above)
Comment (ianwjhalliday) [inline lib/Parser/Scan.cpp]: oh right, my bad, being too hasty

Comment (ianwjhalliday) [inline lib/Parser/Scan.cpp]: and all the more reason it was poor usage

File: test/es6/HTMLComments.js (hunk 3 above)
Comment (tcare) [inline test/es6/HTMLComments.js]: Spoke with Ian offline. I'm exploring the compliance for this manually and not including the ASI tests here. Testing for automatic semicolon insertion in this context is a bit tedious and verbose. What needs to be verified is that a valid HTML end comment gets treated like a newline in an ASI context.

File: lib/Parser/Scan.cpp (hunk 6 above)
Comment (tcare) [inline lib/Parser/Scan.cpp]: They don't need to be there. I'll add a comment to be explicit.

File: test/es6/HTMLComments.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+WScript.Echo("Code before LF--> is reachable");
+--> WScript.Echo("Code after LF--> is unreachable");
+
+// LS u+2028
+WScript.Echo("Code before LS--> is reachable");
--> WScript.Echo("Code after LS--> is unreachable");
+
+// PS u+2029
+WScript.Echo("Code before PS--> is reachable");
--> WScript.Echo("Code after PS--> is unreachable");
```
Comment (dilijev) [inline test/es6/HTMLComments.js]: You already gave the unicode sequences in the comment above, so I think this is redundant.

**Comment by dilijev** [comment]
@dotnet-bot test this please

**Comment by tcare** [comment]
Ping, I'd like to get this merged before EOW

File: lib/Parser/Scan.cpp (hunk 8 above)
Comment (tcare) [inline lib/Parser/Scan.cpp]: Sorry I missed this in my most recent commit. I don't think this is actually allowed in the spec.

**Comment by tcare** [comment]
I missed support for single line case:

/* comment */ -->

Adding that now.

**Comment by ianwjhalliday** [comment]
Hmm, since that case was missed it would be good to add another test file that enumerates all the branches of the grammar productions in simple cases to smoke out any other gotchas if there are any.

**Comment by tcare** [comment]
Yep, I think it's important that we be exhaustive here.

File: test/es6/HTMLComments.js (hunk 11 above)
Comment (tcare) [inline test/es6/HTMLComments.js]: Fixed

**Comment by tcare** [comment]
Ping, could someone please take a look at the update.

File: test/es6/HTMLComments.baseline [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+Code before LF--> is reachable
+Code before LS--> is reachable
+Code before PS--> is reachable
+Code before CRLS--> is reachable
+Code before CRPS--> is reachable
+Code before <!-- is reachable
+Code before <!-- --> is reachable
+Code before <!-- LineTerminator --> is reachable
+Code before /* */ is reachable
+Code before /* */ is reachable
```
Comment (dilijev) [inline test/es6/HTMLComments.baseline]: What about code after `/* */`?

File: test/es6/HTMLComments.js [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+   Line
+   Comment */ --> WScript.Echo("Code after */ --> is unreachable");
+WScript.Echo("Code before /* */ is reachable"); /* Comment */ --> WScript.Echo("Code after /* */ --> is unreachable");
+WScript.Echo("Code before /* */ is reachable"); /* Comment */--> WScript.Echo("Code after /* */--> is unreachable"); // No WhiteSpaceSequence
+
+// Post-decrement with a greater-than comparison does not get interpreted as a comment
+var a = 1; a-->a; WScript.Echo("Code after --> is reachable");
+assert.areEqual(0, a, "Post decrement executes");
+
+assert.throws(function () { eval('/* */ --->'); }, SyntaxError, "HTMLCloseComment causes syntax error with an extra -", "Syntax error");
```
Comment (dilijev) [inline test/es6/HTMLComments.js]: The output message here is misleading because it's not the same as the end comment token when adjacent to a variable. In any other case it shouldn't be reachable. Just asking for a clearer message for the output.

File: test/es6/HTMLComments.baseline (hunk 12 above)
Comment (tcare) [inline test/es6/HTMLComments.baseline]: Sorry that should be --> after the /* */

File: test/es6/HTMLComments.js (hunk 13 above)
Comment (dilijev) [inline test/es6/HTMLComments.js]: Also in general I think it might be better to have the messages about unreachability displayed in the case that the code is unreachable (set a flag before and unset it in the supposedly unreachable code and then print "<incorrect situation> is reachable" if the flag is unset, or something like that).

That way incorrect output in the baseline is more suggestive of the problem.

File: test/es6/module-syntax.js [hunk 14]
```diff
@@ -124,6 +124,16 @@ var tests = [
             assert.doesNotThrow(function () { WScript.LoadModuleFile('.\\module\\ValidReExportStatements.js', 'samethread'); }, "Valid re-export statements");
         }
     },
+    {
+        name: "HTML comments do not parse",
```
Comment (ianwjhalliday) [inline test/es6/module-syntax.js]: nit: "... do not parse in module scripts"

Just to be clear about the intent and scope of this test, that yes, indeed, this is specific to modules.

File: lib/Parser/Scan.cpp [hunk 15]
```diff
@@ -2087,36 +2093,11 @@ tokens Scanner<EncodingPolicy>::ScanCore(bool identifyKwds)
             case '-':
                 p++;
                 token = tkDec;
-                if (m_fHtmlComments)
+                if (!m_fIsModuleCode)
                 {
-                    int i = 0;
-                    while ('-' == PeekFirst(p + i, last)) //Have already seen --, skip any further - characters
-                        i++;
-                    if ('>' == PeekFirst(p + i++, last)) //This means we've got a --------------------------->.
```
Comment (ianwjhalliday) [inline lib/Parser/Scan.cpp]: So the `-------->` is no longer handled, correct?

There should be a test verifying this.

Also, what do the other browsers do about this case?  If they also allow `------>` then we should probably disregard the spec and maintain this behavior for compatibility.

Comment (tcare) [inline lib/Parser/Scan.cpp]: There is a test for this, but GitHub is cutting it off in the diff. You can see it if you view the whole file.

I'll look into what other browsers do.

Comment (ianwjhalliday) [inline lib/Parser/Scan.cpp]: Ah, cool.

File: test/es6/HTMLComments.baseline (hunk 12 above)
Comment (dilijev) [inline test/es6/HTMLComments.baseline]: I assume you mean "that should be 'after the /* */' " and not something involving the HTML comment token `-->` :)

Comment (tcare) [inline test/es6/HTMLComments.baseline]: No, the test is for the single line comment variant /* */ --> which should be the test description.

Comment (dilijev) [inline test/es6/HTMLComments.baseline]: Oh ok, that also makes sense.

File: lib/Parser/Scan.cpp (hunk 15 above)
Comment (tcare) [inline lib/Parser/Scan.cpp]: My comment seemed to get lost. Both Firefox and Chrome reject --->

File: test/es6/HTMLComments.js (hunk 13 above)
Comment (tcare) [inline test/es6/HTMLComments.js]: Fixed the first, wontfix the second. I think it's easy enough to see where something broke for a baseline test.

**Comment by tcare** [comment]
Updated based on feedback from Ian and Doug

File: test/es6/module-syntax.js [hunk 16]
```diff
@@ -124,6 +124,16 @@ var tests = [
             assert.doesNotThrow(function () { WScript.LoadModuleFile('.\\module\\ValidReExportStatements.js', 'samethread'); }, "Valid re-export statements");
         }
     },
+    {
+        name: "HTML comments do not parse in module code,
```
Comment (dilijev) [inline test/es6/module-syntax.js]: There was supposed to be a close quote at the end of the name string, right?

**Comment by ianwjhalliday** [comment]
Looks good

**Comment by tcare** [comment]
@dotnet-bot test this please

**Comment by dilijev** [comment]
LGTM. If you had more time to invest on this I would still push for making the test output indicative of both cases where the code was reachable and cases where it was not, but it's fine the way it is.

**Comment by tcare** [comment]
I think the baseline is fine and it's readable as is.
